website/blog/2022-05-24-joining-snapshot-complexity.md

```diff
@@ -0,0 +1,288 @@
+
+Consider the complexity of the problem: you’ve successfully captured the history of all your tables using snapshots. You have `history_table_1` and `history_table_2`, and you want to join on a common key, `product_id`. However, we cannot just join on the primary key because each table has several rows of history for the same id, all valid across different timespans. 
+
+`history_table_1`: 
+
+| product_id | important_status | dbt_valid_from | dbt_valid_to |
+| --- | --- | --- | --- |
+| 1 | available | 2021-11-15 16:00:0000 | NULL |
+| 1 | pending | 2021-11-10 08:00:000 | 2021-11-15 16:00:0000 |
+| 1 | not_available | 2021-10-01 10:00:000 | 2021-11-10 08:00:000 |
+| 2 | not_available | 2021-11-15 15:30:0000 | NULL  |
+
+`history_table_2`:
+
+| product_id | order_id | product_order_id | order_status | dbt_valid_from | dbt_valid_to |
+| --- | --- | --- | --- | --- | --- |
+| 1 | A | 1A | available | 2021-11-15 16:00:0000 | NULL |
+| 1 | A | 1A | pending | 2021-10-31 12:00:000 | 2021-11-15 16:00:0000 |
+| 1 | B | 1B | available | 2021-11-15 15:30:0000 | NULL |
+| 1 | B | 1B | pending | 2021-11-10 10:00:000 | 2021-11-15 15:30:0000 |
+| 2 | C | 2C | available | 2021-11-10 15:00:0000 | NULL  |
+
+This doesn’t look so bad. How complex can this get? Let’s take a look at the math. Say `historical_table_1` has $x$ historical rows per `product_id`, and $y$ ids total. That’s $x*y = n$ rows of data. `historical_table_2` has $z$  historical rows per `product_id`, and $w$ ids ($z*w = m$ rows). The subsequent join on product_id then [changes the complexity](https://www.freecodecamp.org/news/big-o-notation-why-it-matters-and-why-it-doesnt-1674cfa8a23c/) from $O(n)$ to $O(n*m)$ very quickly ($x*y*z*w$ possibilities!). The complexity continues to increase as we join together more and more historical tables. 
+
+I know what you’re thinking — what a mess! Can’t we just join everything together, and snapshot the resulting table? This is not a bad thought. It would save you the trouble of thinking through a problem with $O(n*m*a*b*c*d*...*q)$ complexity. And in some cases, this may capture all the history you need! 
+
+However, it does not provide a solution to the problem initially posed. The historical records track when each table is valid, rather than when the joined table is valid, and this history for each dataset will only be reflected when you snapshot each table, and then join them, rather than joining and subsequently snapshotting the table. The `valid_from` and `valid_to` built into the joined-then-snapshotted table will only be built from `updated_at` timestamps where the joined table is updated, and thus changes in the underlying data may not be captured. We want to understand when the records are truly valid across all tables, meaning we need to take into account the valid timestamps from each individual dataset. 
```

Comment:
```suggestion
However, it doesn't provide a solution to the problem initially posed. The historical records track when each table is valid, rather than when the joined table is valid, and this history for each dataset will only be reflected when you snapshot each table, and then join them, rather than joining and subsequently snapshotting the table. The `valid_from` and `valid_to` built into the joined-then-snapshotted table will only be built from `updated_at` timestamps where the joined table is updated, and thus changes in the underlying data may not be captured. We want to understand when the records are truly valid across all tables, meaning we need to take into account the valid timestamps from each individual dataset. 
```
To be consistent with contractions